apphub/contrastive_learning/simclr/simclr_tf.py

```diff
@@ -0,0 +1,236 @@
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    x = layers.Add()([x, residual(x, 128)])
+    # layer2
+    x = layers.Conv2D(256, 3, padding='same')(x)
+    x = layers.MaxPool2D()(x)
+    x = layers.BatchNormalization(momentum=0.8)(x)
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    # layer3
+    x = layers.Conv2D(512, 3, padding='same')(x)
+    x = layers.MaxPool2D()(x)
+    x = layers.BatchNormalization(momentum=0.8)(x)
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    x = layers.Add()([x, residual(x, 512)])
+    # layers4
+    x = layers.GlobalMaxPool2D()(x)
+    code = layers.Flatten()(x)
+
+    p_head = layers.Dense(head_len)(code)
+    model_con = tf.keras.Model(inputs=inp, outputs=p_head)
+
+    s_head = layers.Dense(classes)(code)
+    s_head = layers.Activation('softmax', dtype='float32')(s_head)
+    model_finetune = tf.keras.Model(inputs=inp, outputs=s_head)
+
+    return model_con, model_finetune
+
+
+def residual(x, num_channel):
+    """A ResNet unit for ResNet9.
+
+    Args:
+        x: Input Keras tensor.
+        num_channel: The number of layer channel.
+
+    Return:
+        Output Keras tensor.
+    """
+    x = layers.Conv2D(num_channel, 3, padding='same')(x)
+    x = layers.BatchNormalization(momentum=0.8)(x)
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    x = layers.Conv2D(num_channel, 3, padding='same')(x)
+    x = layers.BatchNormalization(momentum=0.8)(x)
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    return x
+
+
+class NTXentOp(TensorOp):
+    def __init__(self, arg1, arg2, outputs, temperature=1.0, mode=None):
+        super().__init__(inputs=(arg1, arg2), outputs=outputs, mode=mode)
+        self.temperature = temperature
+
+    def forward(self, data, state):
+        arg1, arg2 = data
+        loss = NTXent(arg1, arg2, self.temperature)
+        return loss
+
+
+def NTXent(A, B, temperature):
+    large_number = 1e9
+    batch_size = tf.shape(A)[0]
+    A = tf.math.l2_normalize(A, -1)
+    B = tf.math.l2_normalize(B, -1)
+
+    mask = tf.one_hot(tf.range(batch_size), batch_size)
+    labels = tf.one_hot(tf.range(batch_size), 2 * batch_size)
+
+    aa = tf.matmul(A, A, transpose_b=True) / temperature
+    aa = aa - mask * large_number
+    ab = tf.matmul(A, B, transpose_b=True) / temperature
+    bb = tf.matmul(B, B, transpose_b=True) / temperature
+    bb = bb - mask * large_number
+    ba = tf.matmul(B, A, transpose_b=True) / temperature
+    loss_a = tf.nn.softmax_cross_entropy_with_logits(labels, tf.concat([ab, aa], 1))
+    loss_b = tf.nn.softmax_cross_entropy_with_logits(labels, tf.concat([ba, bb], 1))
+    loss = tf.reduce_mean(loss_a + loss_b)
+
+    return loss, ab, labels
+
+
+def pretrain_model(epochs, batch_size, max_train_steps_per_epoch, save_dir):
+    # step 1: prepare dataset
+    train_data, test_data = load_data()
+    pipeline = fe.Pipeline(
+        train_data=train_data,
+        batch_size=batch_size,
+        ops=[
+            ToFloat(inputs="x", outputs="x"),
```

Comment:
`ToFloat` can happen in the last, it can speed up the NumpyOp since those numpyOp will only deal with uint8 instead of float.

Reply:
I benchmark the speed of two pipeline settings.
ToFloat at begin: 24.11 steps/sec 
ToFloat at end:  26.32 steps/sec

It does make pipeline faster though not much. I will update it.